src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/PartitionedRateLimiter.cs

```diff
@@ -0,0 +1,228 @@
+            IEqualityComparer<TKey>? equalityComparer = null)
+        {
+            _limiters = new Dictionary<TKey, Lazy<RateLimiter>>(equalityComparer);
+            _partitioner = partitioner;
+
+            // TODO: Only create timer once there is an active replenishing limiter
+            // TODO: Figure out what interval we should use
+            _timer = new Timer(Replenish, this, TimeSpan.FromSeconds(1), TimeSpan.FromSeconds(1));
+        }
+
+        public override int GetAvailablePermits(TResource resourceID)
+        {
+            return GetRateLimiter(resourceID).GetAvailablePermits();
+        }
+
+        protected override RateLimitLease AcquireCore(TResource resourceID, int permitCount)
+        {
+            return GetRateLimiter(resourceID).Acquire(permitCount);
+        }
+
+        protected override ValueTask<RateLimitLease> WaitAsyncCore(TResource resourceID, int permitCount, CancellationToken cancellationToken)
+        {
+            return GetRateLimiter(resourceID).WaitAsync(permitCount, cancellationToken);
+        }
+
+        private RateLimiter GetRateLimiter(TResource resourceID)
+        {
+            RateLimitPartition<TKey> partition = _partitioner(resourceID);
+            Lazy<RateLimiter>? limiter;
+            lock (Lock)
+            {
+                ThrowIfDisposed();
+                if (!_limiters.TryGetValue(partition.PartitionKey, out limiter))
+                {
+                    // Using Lazy avoids calling user code (partition.Factory) inside the lock
+                    limiter = new Lazy<RateLimiter>(() => partition.Factory(partition.PartitionKey));
+                    _limiters.Add(partition.PartitionKey, limiter);
+                    // Cache is invalid now
+                    _cacheInvalid = true;
+                }
+            }
+            return limiter.Value;
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            if (!disposing)
+            {
+                return;
+            }
+
+            lock (Lock)
+            {
+                if (CommonDisposeUnsynchronized())
+                {
+                    return;
+                }
+
+                foreach (KeyValuePair<TKey, Lazy<RateLimiter>> limiter in _limiters)
+                {
+                    limiter.Value.Value.Dispose();
+                }
+                _limiters.Clear();
+            }
+        }
+
+        protected override async ValueTask DisposeAsyncCore()
+        {
+            Dictionary<TKey, Lazy<RateLimiter>> limiters;
+            lock (Lock)
+            {
+                if (CommonDisposeUnsynchronized())
+                {
+                    return;
+                }
+
+                limiters = _limiters;
+                _limiters = new();
+            }
+
+            foreach (KeyValuePair<TKey, Lazy<RateLimiter>> limiter in limiters)
+            {
+                await limiter.Value.Value.DisposeAsync().ConfigureAwait(false);
```

Comment:
Shouldn't this wait on any pending background calls to `TryReplenish()`?

Reply:
Hmm, probably should. It shouldn't cause issues with the built in limiters, but custom ones might not be as thread-safe. Timer has a nice DisposeAsync method for this, sadly it's netstandard2.1+.

We could call Timer.Dispose with a WaitHandle... but that would be blocking